client/clientutil/modelinfo.go

```diff
@@ -0,0 +1,461 @@
+		// like we do in `snap info`
+		fmt.Fprintf(w, "brand%s\t%s\n", separator, publisher)
+
+		// for model, if there's a display-name, we show that first with the
+		// real model in parenthesis
+		if displayName := modelAssertion.HeaderString("display-name"); displayName != "" {
+			modelHeader = fmt.Sprintf("%s (%s)", displayName, modelHeader)
+		}
+		fmt.Fprintf(w, "model%s\t%s\n", separator, modelHeader)
+	}
+
+	// only output the grade if it is non-empty, either it is not in the model
+	// assertion for all non-uc20 model assertions, or it is non-empty and
+	// required for uc20 model assertions
+	grade := modelAssertion.HeaderString("grade")
+	if grade != "" {
+		fmt.Fprintf(w, "grade%s\t%s\n", separator, grade)
+	}
+
+	storageSafety := modelAssertion.HeaderString("storage-safety")
+	if storageSafety != "" {
+		fmt.Fprintf(w, "storage-safety%s\t%s\n", separator, storageSafety)
+	}
+
+	// serial is same for all variants
+	fmt.Fprintf(w, "serial%s\t%s\n", separator, serial)
+
+	// return if verbose is not requested
+	if !options.Verbose {
+		return w.Flush()
+	}
+	return printVerboseAssertionHeaders(w, options, &modelAssertion)
+}
+
+// PrintModelAssertionYAML will format the provided serial or model assertion based on the parameters given in
+// YAML format. The output will be written to the provided io.Writer.
+func PrintSerialAssertionYAML(w *tabwriter.Writer, modelFormatter ModelFormatter, options PrintModelAssertionOptions, serialAssertion asserts.Serial) error {
+
+	// if we got an invalid model assertion bail early
+	if options.Assertion {
+		_, err := w.Write(asserts.Encode(&serialAssertion))
+		return err
+	}
+
+	// the rest of this function is the main flow for outputting either the
+	// serial assertion in normal or verbose mode
+
+	// ordering of the primary keys for model: brand, model, serial
+	// ordering of primary keys for serial is brand-id, model, serial
+
+	// output brand/brand-id
+	brandIDHeader := serialAssertion.HeaderString("brand-id")
+	modelHeader := serialAssertion.HeaderString("model")
+	serial := serialAssertion.HeaderString("serial")
+
+	fmt.Fprintf(w, "brand-id:\t%s\n", brandIDHeader)
+	fmt.Fprintf(w, "model:\t%s\n", modelHeader)
+	fmt.Fprintf(w, "serial:\t%s\n", serial)
+
+	// return if verbose is not requested
+	if !options.Verbose {
+		return w.Flush()
+	}
+	return printVerboseAssertionHeaders(w, options, &serialAssertion)
+}
+
+// PrintModelAssertionJSON will format the provided serial or model assertion based on the parameters given in
+// JSON format. The output will be written to the provided io.Writer.
+func PrintModelAssertionJSON(w *tabwriter.Writer, modelFormatter ModelFormatter, options PrintModelAssertionOptions, modelAssertion asserts.Model, serialAssertion *asserts.Serial) error {
+
+	serializeJSON := func(v interface{}) error {
+		marshalled, err := json.MarshalIndent(v, "", "  ")
+		if err != nil {
+			return err
+		}
+		_, err = w.Write(marshalled)
+		w.Flush()
+		return err
```

Comment:
If `w.write` succeeds but `w.Flush()` fails you would still return `nil` as error here.

Reply:
Nice catch, updated this